internal/wasm/jit/engine.go

```diff
@@ -17,17 +18,33 @@ import (
 type (
 	// engine is an JIT implementation of internalwasm.Engine
 	engine struct {
+		// compiledFunctions are the currently compiled functions.
+		// The index means wasm.FunctionAddress, but we intentionally avoid using map
+		// as the underlying memory region is accessed by assembly directly by
+		// using compiledFunctionsElement0Address.
+		compiledFunctions []*compiledFunction
+
+		// mux is used for read/write access to compiledFunctions slice.
+		// This is necessary as each compiled function will access the slice from the native code
+		// when they make function calls while engine might be modifying the underlying slice when
+		// adding a new compiled function. We take read lock when creating new virtualMachine
+		// for each function invocation while take write lock in engine.addCompiledFunction.
+		mux sync.RWMutex
```

Comment:
this can be removed by assuming that `engine.Compile` would never be called from a different goroutine, but I think sooner or later we would add. That's why I added, but I'm fine removing this for now